src/control_flow/mod.rs

```diff
@@ -0,0 +1,793 @@
+          self.mark_as_end(lo, end);
+          match end {
+            e if matches!(e, End::Forced { .. }) => {
+              self.scope.end = Some(e);
+            }
+            _ => {
+              self.scope.end = prev_end;
+            }
+          }
+        }
+      }
+    }
+  }
+
+  fn get_end_reason(&self, lo: BytePos) -> Option<End> {
+    self.info.get(&lo).map(|md| md.end).flatten()
+  }
+
+  /// Mark a statement as finisher - finishes execution - and expose it.
+  fn mark_as_end(&mut self, lo: BytePos, end: End) {
+    let new_end = match self.scope.end {
+      // `End::Continue` doesn't mean much about execution status, just indicating that execution has
+      // not yet stopped so far. So if `End::Forced` or `End::Break` comes and the current
+      // `self.scope.end` is `Some(End::Continue)`, then `self.scope.end` should be replaced with the
+      // coming value.
+      None | Some(End::Continue) => {
+        self.scope.end = Some(end);
+        Some(end)
+      }
+      Some(End::Break) => Some(end),
+      Some(e) => e.merge_forced(end).or(self.scope.end),
+    };
+
+    self.info.entry(lo).or_default().end = new_end;
+  }
+
+  /// Visits statement or block. This method handles break and continue.
+  ///
+  /// This cannot be done in visit_stmt of Visit because
+  ///  this operation is very opinionated.
+  fn visit_stmt_or_block(&mut self, s: &Stmt) {
+    s.visit_with(&Invalid { span: DUMMY_SP }, self);
+
+    // break, continue **may** make execution end
+    match s {
+      Stmt::Break(..) | Stmt::Continue(..) => {
+        self.mark_as_end(s.span().lo, End::Break)
+      }
+      _ => {}
+    }
+  }
+}
+
+impl Visit for Analyzer<'_> {
+  noop_visit_type!();
+
+  fn visit_return_stmt(&mut self, n: &ReturnStmt, _: &dyn Node) {
+    n.visit_children_with(self);
+    self.mark_as_end(n.span().lo, End::forced_return());
+  }
+
+  fn visit_throw_stmt(&mut self, n: &ThrowStmt, _: &dyn Node) {
+    n.visit_children_with(self);
+    self.mark_as_end(n.span().lo, End::forced_throw());
+  }
+
+  fn visit_break_stmt(&mut self, n: &BreakStmt, _: &dyn Node) {
+    if let Some(label) = &n.label {
+      let label = label.to_id();
+      self.scope.found_break = Some(Some(label));
+    } else {
+      self.scope.found_break = Some(None);
+    }
+  }
+
+  fn visit_continue_stmt(&mut self, _: &ContinueStmt, _: &dyn Node) {
+    self.scope.found_continue = true;
+  }
+
+  fn visit_block_stmt(&mut self, s: &BlockStmt, _: &dyn Node) {
+    s.visit_children_with(self);
+
+    if let Some(end) = self.scope.end {
+      self.mark_as_end(s.span.lo, end);
+    } else {
+      self.mark_as_end(s.span.lo, End::Continue);
+    }
+  }
+
+  fn visit_stmts(&mut self, stmts: &[Stmt], _: &dyn Node) {
+    for stmt in stmts {
+      self.visit_stmt_or_block(stmt);
+    }
+  }
+
+  fn visit_expr(&mut self, n: &Expr, _: &dyn Node) {
+    n.visit_children_with(self);
+
+    if matches!(self.scope.end, None | Some(End::Continue)) {
```

Comment:
The only change for the implementation is here.
Previously this line looked like `if self.scope.end.is_none() {`. Actually we should set `self.scope.may_throw` to `true` in this case too.